pkg/addonfactory/helmaddonfactory/helmagentaddon.go

```diff
@@ -0,0 +1,155 @@
+
+// the build-in values
+const (
+	clusterName           string = "clusterName"
+	addonInstallNamespace string = "addonInstallNamespace"
+	hubKubeConfigSecret   string = "hubKubeConfigSecret"
+)
+
+type Values map[string]interface{}
+
+type getValuesFunc func(cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) (Values, error)
+
+type HelmAgentAddon struct {
+	decoder           runtime.Decoder
+	chart             *chart.Chart
+	getValuesFunc     getValuesFunc
+	agentAddonOptions agent.AgentAddonOptions
+}
+
+func newHelmAgentAddon(
+	scheme *runtime.Scheme,
+	chart *chart.Chart,
+	getValuesFunc getValuesFunc,
+	agentAddonOptions agent.AgentAddonOptions) *HelmAgentAddon {
+	return &HelmAgentAddon{
+		decoder:           serializer.NewCodecFactory(scheme).UniversalDeserializer(),
+		chart:             chart,
+		getValuesFunc:     getValuesFunc,
+		agentAddonOptions: agentAddonOptions}
+}
+
+func (a *HelmAgentAddon) Manifests(
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) ([]runtime.Object, error) {
+	var objects []runtime.Object
+
+	values, err := a.getValues(cluster, addon)
+	if err != nil {
+		return objects, err
+	}
+
+	helmEngine := engine.Engine{
+		Strict:   true,
+		LintMode: false,
+	}
+
+	templates, err := helmEngine.Render(a.chart, values)
+	if err != nil {
+		return objects, err
+	}
+	for _, data := range templates {
+		if len(data) == 0 {
+			continue
+		}
+		fmt.Printf("%s \n", data)
+		object, _, err := a.decoder.Decode([]byte(data), nil, nil)
+		if err != nil {
+			return nil, err
+		}
+		objects = append(objects, object)
+	}
+
+	return objects, nil
+}
+
+func (a *HelmAgentAddon) GetAgentAddonOptions() agent.AgentAddonOptions {
+	return a.agentAddonOptions
+}
+
+func (a *HelmAgentAddon) getValues(
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) (chartutil.Values, error) {
+	overrideValues := map[string]interface{}{}
+
+	if a.getValuesFunc != nil {
+		userValues, err := a.getValuesFunc(cluster, addon)
+		if err != nil {
+			return overrideValues, err
+		}
+		overrideValues = userValues
+	}
+	overrideValues = MergeValues(overrideValues, a.getBuildInValues(cluster, addon))
+
+	values, err := chartutil.ToRenderValues(a.chart, overrideValues,
+		a.releaseOptions(cluster, addon), a.capabilities(cluster, addon))
+	if err != nil {
+		klog.Error("failed to render helm chart with values %v. err:%v", overrideValues, err)
+		return values, err
+	}
+
+	return values, nil
+}
+
+func (a *HelmAgentAddon) getBuildInValues(
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) Values {
+	buildInValues := Values{}
+
+	buildInValues[clusterName] = cluster.GetName()
+
+	installNamespace := addon.Spec.InstallNamespace
+	if len(installNamespace) == 0 {
+		installNamespace = AddonDefaultInstallNamespace
+	}
+	buildInValues[addonInstallNamespace] = installNamespace
+
+	if a.agentAddonOptions.Registration != nil {
```

Comment:
Add a comment and todo here. It also depdends on the signer configuration in registration, and the registration is an array actually

Reply:
added a TODO here